doc: document json rpc endpoints (#27225)
bitcoin/bitcoin: doc/JSON-RPC-interface.md

option. In the GUI it is possible to execute RPC methods in the Debug Console
Dialog.

## Endpoints

There are two JSON-RPC endpoints on the server:

1. `/`

> 2. `/wallet/<walletname>/`

Reviewer: A few ideas about these:
- could include full path `http://localhost:{port}/`
- could list relevant ports per network (i.e. main=8332, regtest=18443 etc)
- not sure if this applies here but I'm used to seeing API endpoints with variables expressed like `/wallet/:walletname`, at least for REST APIs. I dunno if there is otherwise a convention for that

Author: Personally I feel it's clearer just to use the endpoints themselves, rather than cluttering with hosts and ports.

Also, we have examples now!

### `/` endpoint

This endpoint is always active and can service both wallet and non-wallet

> requests when there are zero or one wallets loaded.

Author: I see your point here, but I was also trying to distinguish: active=enabled, services=wallet/non-wallet.

How about:

> This endpoint is always active. It can always service non-wallet requests and can service wallet requests when exactly one wallet is loaded.

> requests when there are zero or one wallets loaded.

Author: Done

### `/wallet/<walletname>/` endpoint

> This endpoint is only activated when the wallet component has been compiled in

Reviewer: It is also always active for non-wallet requests before any wallets are loaded.

Author: (but only when the wallet is compiled in).

Also I think i've covered that it handles both types of request when no wallet is loaded in the Quirks section?

and can service both wallet and non-wallet requests. It MUST be used for wallet
requests when two or more wallets are loaded.

> This endpoint is equivalent to using bitcoin-cli with the `-rpcwallet=` parameter.

Reviewer: ```suggestion
This is the endpoint used by bitcoin-cli when a `-rpcwallet=` parameter is passed in.
```

Author: Agree with this.

Author: Done

Best practice would dictate using the `/wallet/<walletname>/` endpoint for ALL

> requests when multiple wallets are in use.

Author: Oh, I didn't know you could use the endpoint without a wallet name at all!

IMO it makes sense to specify that it should be used with the `wallet/<walletname>/` syntax for multiwallet (as designed?)

> requests when multiple wallets are in use.

Reviewer: yeah i dunno whats best to reccomend. I think personally Id use `/` for node requests and `/wallet/walletname` for any wallet request no matter how many wallets are loaded.

> requests when multiple wallets are in use.

Author: Leaving this for now, as using the bare `/wallet/` endpoint could be considered unsupported behaviour.

### Quirks:

* With more than one wallet loaded the `/` endpoint will still be able to
service non-wallet requests.
* When active the `/wallet/<walletname>/` endpoint can still be used when no

> wallet is loaded by providing a dummy `walletname`.

Reviewer: I think these can be folded in above if you like my suggestions.

I think a few `curl` examples might be appropriate here as well since this document is covering raw endpoint stuff

Author: Again I think I agree here, especially seeing as there are no /wallet/ examples given in RPCHelp

Author: Removed quirks

option. In the GUI it is possible to execute RPC methods in the Debug Console
Dialog.

## Endpoints

There are two JSON-RPC endpoints on the server:

1. `/`
2. `/wallet/<walletname>/`

### `/` endpoint

This endpoint is always active.
It can always service non-wallet requests and can service wallet requests when
exactly one wallet is loaded.

### `/wallet/<walletname>/` endpoint

This endpoint is only activated when the wallet component has been compiled in.
It can service both wallet and non-wallet requests.
It MUST be used for wallet requests when two or more wallets are loaded.

This is the endpoint used by bitcoin-cli when a `-rpcwallet=` parameter is passed in.

Best practice would dictate using the `/wallet/<walletname>/` endpoint for ALL
requests when multiple wallets are in use.

### Examples

```sh
# Get block count from the / endpoint when rpcuser=alice and rpcport=38332
$ curl --user alice --data-binary '{"jsonrpc": "1.0", "id": "0", "method": "getblockcount", "params": []}' -H 'content-type: text/plain;' localhost:38332/

# Get balance from the /wallet/walletname endpoint when rpcuser=alice, rpcport=38332 and rpcwallet=desc-wallet

> $ curl --user alice --data-binary '{"jsonrpc": "1.0", "id": "0", "method": "getbalance", "params": []}' -H 'content-type: text/plain;' localhost:38332/wallet/desc-wallet

Reviewer 2: Why are the examples using signet port? I think we should use mainnet port (8332) same as all the [RPC examples](https://github.com/bitcoin/bitcoin/blob/23e2bfcbc42849daa8e2b69f9bbdc526bc8743a7/src/rpc/util.cpp#L160).

Author: I'm happy to be persuaded either way here, no particular reason for using signet (other than I was using it for testing). Although I personally don't think it matters much as the comment for each shows which variables are being used
